src/libPMacc/include/particles/ParticlesBase.kernel

```diff
@@ -755,6 +459,324 @@ struct KernelFillGaps
+                            destFrames[ linearIdx ] = tmpFrame;
+                            pb.setAsLastFrame(
+                                tmpFrame,
+                                relativeCtx[ idx ]
+                            );
+                        }
+                        /* check if a `high frame` is needed */
+                        if ( destFramesCounter[ linearIdx ] > frameSize )
+                        {
+                            lastFrameSizeCtx[ idx ] = destFramesCounter[ linearIdx ] - frameSize;
+                            FramePtr tmpFrame( pb.getEmptyFrame( ) );
+                            destFrames[ linearIdx + numExchanges ] = tmpFrame;
+                            pb.setAsLastFrame(
+                                tmpFrame,
+                                relativeCtx[ idx ]
+                            );
+                        }
+                    }
+                }
+            );
+            __syncthreads( );
+
+            forEachParticle(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const idx
+                )
+                {
+                    /* All threads with a valid index in the neighbor's frame, valid index
+                     * range is [0, frameSize * 2-1], will copy their particle to the new
+                     * frame.
+                     *
+                     * The default value for indexes (in the destination frame) is
+                     * above this range (INV_LOC_IDX) for all particles that are not shifted.
+                     */
+                    if ( destParticleIdxCtx[ idx ] < frameSize * 2 )
+                    {
+                        if ( destParticleIdxCtx[ idx ] >= frameSize )
+                        {
+                            /* use `high frame` */
+                            directionCtx[ idx ] += numExchanges;
+                            destParticleIdxCtx[ idx ] -= frameSize;
+                        }
+                        auto dstParticle = destFrames[ directionCtx[ idx ] ][ destParticleIdxCtx[ idx ] ];
+                        auto srcParticle = frame[ linearIdx ];
+                        dstParticle[ multiMask_ ] = 1;
+                        srcParticle[ multiMask_ ] = 0;
+                        auto dstFilteredParticle =
+                            particles::operations::deselect< multiMask >( dstParticle );
+                        particles::operations::assign(
+                            dstFilteredParticle,
+                            srcParticle
+                        );
+                    }
+                }
+            );
+            __syncthreads( );
+
+            forEachExchange(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    /* if the `low frame` is now full, each master thread removes it and
```

Comment:
I think the `low frame` is not really *removed* but just not managed any more by the master thread (and a new one taken instead), right?